File: lib/models/wearable_data.dart
class WearableData {
  final String id;
  final String patientId;
  final String deviceId;
  final WearableDataType dataType;
  final DateTime timestamp;
  final DateTime hourlyTimestamp;
  final double value;
  final double? secondaryValue;
  final String unit;
  final int quality;
  final WearableDataStatus status;
  final WearableDataMetadata? metadata;
  final bool isProcessed;
  final bool isAggregated;
  final bool triggerAlert;
  final WearableAlertLevel alertLevel;
  final WearableDataSource source;
  final String? syncBatchId;
  final DateTime createdAt;
  final DateTime updatedAt;

  const WearableData({
    required this.id,
    required this.patientId,
    required this.deviceId,
    required this.dataType,
    required this.timestamp,
    required this.hourlyTimestamp,
    required this.value,
    this.secondaryValue,
    required this.unit,
    this.quality = 100,
    required this.status,
    this.metadata,
    this.isProcessed = false,
    this.isAggregated = false,
    this.triggerAlert = false,
    this.alertLevel = WearableAlertLevel.none,
    this.source = WearableDataSource.realTime,
    this.syncBatchId,
    required this.createdAt,
    required this.updatedAt,
  });

  factory WearableData.fromJson(Map<String, dynamic> json) {
    return WearableData(
      id: json['_id'] ?? json['id'] ?? '',
      patientId: json['patientId'] ?? '',
      deviceId: json['deviceId'] ?? '',
      dataType: WearableDataType.fromString(json['dataType'] ?? ''),
      timestamp: DateTime.parse(json['timestamp'] ?? DateTime.now().toIso8601String()),
      hourlyTimestamp: DateTime.parse(json['hourlyTimestamp'] ?? DateTime.now().toIso8601String()),
      value: (json['value'] ?? 0).toDouble(),
      secondaryValue: json['secondaryValue']?.toDouble(),
      unit: json['unit'] ?? '',
      quality: json['quality'] ?? 100,
      status: WearableDataStatus.fromString(json['status'] ?? 'normal'),
      metadata: json['metadata'] != null ? WearableDataMetadata.fromJson(json['metadata']) : null,
      isProcessed: json['isProcessed'] ?? false,
      isAggregated: json['isAggregated'] ?? false,
      triggerAlert: json['triggerAlert'] ?? false,
      alertLevel: WearableAlertLevel.fromString(json['alertLevel'] ?? 'none'),
      source: WearableDataSource.fromString(json['source'] ?? 'real_time'),
      syncBatchId: json['syncBatchId'],
      createdAt: DateTime.parse(json['createdAt'] ?? DateTime.now().toIso8601String()),
      updatedAt: DateTime.parse(json['updatedAt'] ?? DateTime.now().toIso8601String()),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'patientId': patientId,
      'deviceId': deviceId,
      'dataType': dataType.value,
      'timestamp': timestamp.toIso8601String(),
      'hourlyTimestamp': hourlyTimestamp.toIso8601String(),
      'value': value,
      'secondaryValue': secondaryValue,
      'unit': unit,
      'quality': quality,
      'status': status.value,
      'metadata': metadata?.toJson(),
      'isProcessed': isProcessed,
      'isAggregated': isAggregated,
      'triggerAlert': triggerAlert,
      'alertLevel': alertLevel.value,
      'source': source.value,
      'syncBatchId': syncBatchId,
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt.toIso8601String(),
    };
  }

  String get formattedValue {
    if (secondaryValue != null) {
      return '${value.toInt()}/${secondaryValue!.toInt()} $unit';
    }
    return '$value $unit';
  }

  String get timeAgo {
    final now = DateTime.now();
    final difference = now.difference(timestamp);
    
    if (difference.inDays > 0) return '${difference.inDays}d ago';
    if (difference.inHours > 0) return '${difference.inHours}h ago';
    if (difference.inMinutes > 0) return '${difference.inMinutes}m ago';
    return 'Just now';
  }

  bool get isRecent {
    final fiveMinutesAgo = DateTime.now().subtract(const Duration(minutes: 5));
    return timestamp.isAfter(fiveMinutesAgo);
  }

  bool get isWithinNormalRange {
    return status == WearableDataStatus.normal;
  }
}

class WearableDataMetadata {
  final String? activityType;
  final String? sleepStage;
  final int? ecgDuration;
  final int? ecgSampleRate;
  final String? stressLevel;
  final double? ambientTemperature;
  final double? humidity;
  final int? batteryLevel;
  final int? signalStrength;

  const WearableDataMetadata({
    this.activityType,
    this.sleepStage,
    this.ecgDuration,
    this.ecgSampleRate,
    this.stressLevel,
    this.ambientTemperature,
    this.humidity,
    this.batteryLevel,
    this.signalStrength,
  });

  factory WearableDataMetadata.fromJson(Map<String, dynamic> json) {
    return WearableDataMetadata(
      activityType: json['activityType'],
      sleepStage: json['sleepStage'],
      ecgDuration: json['ecgDuration'],
      ecgSampleRate: json['ecgSampleRate'],
      stressLevel: json['stressLevel'],
      ambientTemperature: json['ambientTemperature']?.toDouble(),
      humidity: json['humidity']?.toDouble(),
      batteryLevel: json['batteryLevel'],
      signalStrength: json['signalStrength'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'activityType': activityType,
      'sleepStage': sleepStage,
      'ecgDuration': ecgDuration,
      'ecgSampleRate': ecgSampleRate,
      'stressLevel': stressLevel,
      'ambientTemperature': ambientTemperature,
      'humidity': humidity,
      'batteryLevel': batteryLevel,
      'signalStrength': signalStrength,
    };
  }
}

enum WearableDataType {
  heartRate('heart_rate'),
  bloodPressure('blood_pressure'),
  oxygenSaturation('oxygen_saturation'),
  temperature('temperature'),
  steps('steps'),
  calories('calories'),
  sleep('sleep'),
  ecg('ecg'),
  stress('stress'),
  activity('activity');

  const WearableDataType(this.value);
  final String value;

  static WearableDataType fromString(String value) {
    for (var type in WearableDataType.values) {
      if (type.value == value) return type;
    }
    return WearableDataType.heartRate;
  }

  String get displayName {
    switch (this) {
      case WearableDataType.heartRate:
        return 'Heart Rate';
      case WearableDataType.bloodPressure:
        return 'Blood Pressure';
      case WearableDataType.oxygenSaturation:
        return 'Oxygen Saturation';
      case WearableDataType.temperature:
        return 'Temperature';
      case WearableDataType.steps:
        return 'Steps';
      case WearableDataType.calories:
        return 'Calories';
      case WearableDataType.sleep:
        return 'Sleep';
      case WearableDataType.ecg:
        return 'ECG';
      case WearableDataType.stress:
        return 'Stress';
      case WearableDataType.activity:
        return 'Activity';
    }
  }

  String get defaultUnit {
    switch (this) {
      case WearableDataType.heartRate:
        return 'bpm';
      case WearableDataType.bloodPressure:
        return 'mmHg';
      case WearableDataType.oxygenSaturation:
        return '%';
      case WearableDataType.temperature:
        return '°C';
      case WearableDataType.steps:
        return 'steps';
      case WearableDataType.calories:
        return 'kcal';
      case WearableDataType.sleep:
        return 'hours';
      case WearableDataType.ecg:
        return 'mV';
      case WearableDataType.stress:
        return 'level';
      case WearableDataType.activity:
        return 'minutes';
    }
  }
}

enum WearableDataStatus {
  normal('normal'),
  warning('warning'),
  critical('critical'),
  unknown('unknown');

  const WearableDataStatus(this.value);
  final String value;

  static WearableDataStatus fromString(String value) {
    for (var status in WearableDataStatus.values) {
      if (status.value == value) return status;
    }
    return WearableDataStatus.unknown;
  }

  String get displayName {
    switch (this) {
      case WearableDataStatus.normal:
        return 'Normal';
      case WearableDataStatus.warning:
        return 'Warning';
      case WearableDataStatus.critical:
        return 'Critical';
      case WearableDataStatus.unknown:
        return 'Unknown';
    }
  }
}

enum WearableAlertLevel {
  none('none'),
  low('low'),
  medium('medium'),
  high('high'),
  critical('critical');

  const WearableAlertLevel(this.value);
  final String value;

  static WearableAlertLevel fromString(String value) {
    for (var level in WearableAlertLevel.values) {
      if (level.value == value) return level;
    }
    return WearableAlertLevel.none;
  }

  String get displayName {
    switch (this) {
      case WearableAlertLevel.none:
        return 'None';
      case WearableAlertLevel.low:
        return 'Low';
      case WearableAlertLevel.medium:
        return 'Medium';
      case WearableAlertLevel.high:
        return 'High';
      case WearableAlertLevel.critical:
        return 'Critical';
    }
  }
}

enum WearableDataSource {
  realTime('real_time'),
  batchSync('batch_sync'),
  manualEntry('manual_entry');

  const WearableDataSource(this.value);
  final String value;

  static WearableDataSource fromString(String value) {
    for (var source in WearableDataSource.values) {
      if (source.value == value) return source;
    }
    return WearableDataSource.realTime;
  }

  String get displayName {
    switch (this) {
      case WearableDataSource.realTime:
        return 'Real-time';
      case WearableDataSource.batchSync:
        return 'Batch Sync';
      case WearableDataSource.manualEntry:
        return 'Manual Entry';
    }
  }
}

File: backend/src/models/WearableData.js
const mongoose = require('mongoose');

const WearableDataSchema = new mongoose.Schema({
  patientId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Patient',
    required: true
  },
  deviceId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'WearableDevice',
    required: true
  },
  dataType: {
    type: String,
    enum: ['heart_rate', 'blood_pressure', 'oxygen_saturation', 'temperature', 'steps', 'calories', 'sleep', 'ecg', 'stress', 'activity'],
    required: true
  },
  timestamp: {
    type: Date,
    required: true,
    default: Date.now
  },
  // For hourly aggregated data
  hourlyTimestamp: {
    type: Date,
    required: true
  },
  // Main value (e.g., heart rate, steps count, temperature)
  value: {
    type: Number,
    required: true
  },
  // Secondary value (e.g., diastolic BP when value is systolic)
  secondaryValue: {
    type: Number
  },
  unit: {
    type: String,
    required: true,
    trim: true
  },
  // Quality indicator (0-100)
  quality: {
    type: Number,
    min: 0,
    max: 100,
    default: 100
  },
  // Status based on normal ranges
  status: {
    type: String,
    enum: ['normal', 'warning', 'critical', 'unknown'],
    default: 'normal'
  },
  // Additional metadata
  metadata: {
    // For activity data
    activityType: {
      type: String,
      enum: ['rest', 'light', 'moderate', 'vigorous', 'sleep']
    },
    // For sleep data
    sleepStage: {
      type: String,
      enum: ['awake', 'light', 'deep', 'rem']
    },
    // For ECG data
    ecgDuration: Number,
    ecgSampleRate: Number,
    // For stress data
    stressLevel: {
      type: String,
      enum: ['low', 'medium', 'high']
    },
    // Environmental factors
    ambientTemperature: Number,
    humidity: Number,
    // Device specific
    batteryLevel: Number,
    signalStrength: Number
  },
  // Flags for data processing
  isProcessed: {
    type: Boolean,
    default: false
  },
  isAggregated: {
    type: Boolean,
    default: false
  },
  // Alert flags
  triggerAlert: {
    type: Boolean,
    default: false
  },
  alertLevel: {
    type: String,
    enum: ['none', 'low', 'medium', 'high', 'critical'],
    default: 'none'
  },
  // Data source tracking
  source: {
    type: String,
    enum: ['real_time', 'batch_sync', 'manual_entry'],
    default: 'real_time'
  },
  syncBatchId: {
    type: String,
    trim: true
  }
}, {
  timestamps: true,
  toJSON: { virtuals: true },
  toObject: { virtuals: true }
});

// Compound indexes for efficient querying
WearableDataSchema.index({ patientId: 1, dataType: 1, hourlyTimestamp: -1 });
WearableDataSchema.index({ deviceId: 1, timestamp: -1 });
WearableDataSchema.index({ hourlyTimestamp: -1, dataType: 1 });
WearableDataSchema.index({ patientId: 1, timestamp: -1 });
WearableDataSchema.index({ triggerAlert: 1, alertLevel: 1 });
WearableDataSchema.index({ isProcessed: 1, timestamp: 1 });

// TTL index for automatic data cleanup (optional - keep 1 year)
WearableDataSchema.index({ timestamp: 1 }, { expireAfterSeconds: 31536000 });

// Virtual for formatted value display
WearableDataSchema.virtual('formattedValue').get(function() {
  if (this.secondaryValue !== undefined) {
    return `${this.value}/${this.secondaryValue} ${this.unit}`;
  }
  return `${this.value} ${this.unit}`;
});

// Virtual for time ago
WearableDataSchema.virtual('timeAgo').get(function() {
  const now = new Date();
  const diff = now - this.timestamp;
  const minutes = Math.floor(diff / 60000);
  const hours = Math.floor(minutes / 60);
  const days = Math.floor(hours / 24);
  
  if (days > 0) return `${days}d ago`;
  if (hours > 0) return `${hours}h ago`;
  if (minutes > 0) return `${minutes}m ago`;
  return 'Just now';
});

// Method to determine if value is within normal range
WearableDataSchema.methods.checkNormalRange = function() {
  const normalRanges = {
    heart_rate: { min: 60, max: 100 },
    oxygen_saturation: { min: 95, max: 100 },
    temperature: { min: 36.1, max: 37.2 },
    blood_pressure: { 
      systolic: { min: 90, max: 140 },
      diastolic: { min: 60, max: 90 }
    }
  };

  const range = normalRanges[this.dataType];
  if (!range) return 'unknown';

  if (this.dataType === 'blood_pressure') {
    const systolicNormal = this.value >= range.systolic.min && this.value <= range.systolic.max;
    const diastolicNormal = this.secondaryValue >= range.diastolic.min && this.secondaryValue <= range.diastolic.max;
    
    if (systolicNormal && diastolicNormal) return 'normal';
    if (this.value > range.systolic.max || this.secondaryValue > range.diastolic.max) return 'critical';
    return 'warning';
  } else {
    if (this.value >= range.min && this.value <= range.max) return 'normal';
    if (this.value > range.max * 1.2 || this.value < range.min * 0.8) return 'critical';
    return 'warning';
  }
};

// Pre-save middleware to set hourly timestamp and status
WearableDataSchema.pre('save', function(next) {
  // Set hourly timestamp for aggregation
  const hour = new Date(this.timestamp);
  hour.setMinutes(0, 0, 0);
  this.hourlyTimestamp = hour;

  // Determine status based on value
  this.status = this.checkNormalRange();

  // Set alert flags for critical values
  if (this.status === 'critical') {
    this.triggerAlert = true;
    this.alertLevel = 'critical';
  } else if (this.status === 'warning') {
    this.triggerAlert = true;
    this.alertLevel = 'medium';
  }

  next();
});

// Static method to get latest data by type
WearableDataSchema.statics.getLatestByType = function(patientId, dataType, limit = 10) {
  return this.find({ patientId, dataType })
    .sort({ timestamp: -1 })
    .limit(limit)
    .populate('deviceId', 'deviceName deviceType');
};

// Static method to get hourly aggregated data
WearableDataSchema.statics.getHourlyData = function(patientId, dataType, startDate, endDate) {
  return this.aggregate([
    {
      $match: {
        patientId: new mongoose.Types.ObjectId(patientId),
        dataType,
        hourlyTimestamp: {
          $gte: startDate,
          $lte: endDate
        }
      }
    },
    {
      $group: {
        _id: '$hourlyTimestamp',
        avgValue: { $avg: '$value' },
        minValue: { $min: '$value' },
        maxValue: { $max: '$value' },
        count: { $sum: 1 },
        criticalCount: {
          $sum: { $cond: [{ $eq: ['$status', 'critical'] }, 1, 0] }
        }
      }
    },
    {
      $sort: { '_id': 1 }
    }
  ]);
};

// Static method to get alert data
WearableDataSchema.statics.getAlerts = function(patientId, alertLevel, limit = 50) {
  const match = { patientId, triggerAlert: true };
  if (alertLevel) match.alertLevel = alertLevel;

  return this.find(match)
    .sort({ timestamp: -1 })
    .limit(limit)
    .populate('deviceId', 'deviceName deviceType');
};

module.exports = mongoose.model('WearableData', WearableDataSchema);
